scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -588,19 +630,172 @@ def __init__(self, pid, monitor_id, logger):
+            # other errors like this process does not have permission to send
+            # a signal to self.pid.  But, if that error is returned to us, we
+            # know the process is running at least, so we ignore the error.
+            return e.errno != 3
+
+    def collect(self):
+        """
+        Collects the metrics from the gathers
+        """
+
+        collector = {}
+        for gather in self.gathers:
+            try:
+                collector.update(gather.run_single_cycle(collector=collector))
+            except Exception as ex:
+                self._logger.exception(
+                    'Exception while collecting metrics for PID: %s of type: %s. Details: %s',
+                    self.pid, type(gather), repr(ex)
+                )
+        return collector
+
+
+class ProcessList(object):
+    """
+    Class that encapsulates the listing of process(es) based on IDs, regular expressions etc.
+    """
+
+    def __init__(self):
+        # list of {'pid': processid, 'ppid': parentprocessid,  'cmd': command }
+        self.processes = []
+        # key -> parent process id, value -> [child process ids...]
+        self.parent_to_children_map = defaultdict(list)
+
+        cmd = ['ps', 'axo', 'pid,ppid,command']
+        sub_proc = Popen(cmd, shell=False, stdout=PIPE)
+        regex = r"\s+(\d+)\s+(\d+)\s+(.*)"
+        lines = sub_proc.stdout.readlines()
+        for line in lines:
+            match = re.search(regex, line)
+            if match:
+                _pid, _ppid, _cmd = match.groups()
+                self.processes.append({'pid': int(_pid), 'ppid': int(_ppid), 'cmd': _cmd})
+
+        self._init_parent_to_child_map()
+
+    def _init_parent_to_child_map(self):
+        for _process in self.processes:
+            self.parent_to_children_map[_process['ppid']].append(_process['pid'])
+
+    def get_matches_commandline(self, match_pattern):
+        """
+        Given a string, match the processes on the name
+        @param match_pattern: process command match pattern
+        @return: List of Process IDs
+        """
+
+        matches = []
+        for _process in self.processes:
+            if re.search(match_pattern, _process['cmd']):
+                matches.append(_process['pid'])
+        return matches
+
+    def get_child_processes(self, ppid):
+        """
+        Given a process pid, get the children pids
+        @param ppid: parent process id
+        @return: list of children process IDs
+        """
+
+        all_children = []
+        children_to_explore = []
+        for _pid in self.parent_to_children_map[ppid]:
+            all_children.append(_pid)
+            children_to_explore.append(_pid)
+
+        # get the children 'recursively'
+        while children_to_explore:  # the invariant
+            child_to_explore = children_to_explore.pop()
+            all_children.extend(self.parent_to_children_map[child_to_explore])
```

Comment:
To do true recursive, I think we must invoke `get_child_process` here.  After all, don't we want the children of the children as well?

Reply:
I was avoiding the recursive solution, but you're right - there is a bug. I forgot to iterate over the children and add them to the list. Will make the change and add the tests.

```
def get_child_processes(ppid, node):
    all_children = []
    children_to_explore = set()
    for _pid in parent_to_children_map[ppid]:
        all_children.append(_pid)
        children_to_explore.add(_pid)

    # get the children 'recursively'
    while children_to_explore:  # the invariant
        child_to_explore = children_to_explore.pop()
        if not parent_to_children_map.get(child_to_explore):
            continue
        unvisited = parent_to_children_map[child_to_explore]
        for node in unvisited:
            children_to_explore.add(node)
            all_children.append(node)
    return list(set(all_children))

if __name__ == '__main__':

    # p1 -> p2 -> p4 -> p5
    #        |
    #        p3 

    # p6 -> p7

    parent_to_children_map = {'p1': ['p2'], 'p2': ['p3', 'p4'], 'p4': ['p5'], 'p6': ['p7']}
    print get_child_processes('p1', parent_to_children_map)
```

this returns the recursed list of children:
```
['p2', 'p3', 'p4', 'p5']`
```